### ISSUE TITLE
Template ads integration for Dianomi

### ISSUE THREAD

**Issue by dianomi-mjs** [ampproject/amphtml#19951]
## What's the issue?

We have been serving AMP ads via Firebolt from Cloudflare and they've told us that they are closing this service down.  So we're looking to sign them ourselves.

## How do we reproduce the issue?

Here's an example of how we are integrated

On:

https://amp-kiplinger-com.cdn.ampproject.org/v/amp.kiplinger.com/slideshow/retirement/T006-S001-worst-states-for-retirement-2018.html?amp_js_v=a2&amp_gsa=1&usqp=mq331AQECAFYAQ%3D%3D

You can see

<amp-ad class="i-amphtml-layout-fixed i-amphtml-layout-size-defined" data-cf-network=dianomi data-id=3186 height=100 i-amphtml-layout=fixed src="https://www.dianomi.com/smartads.pl?format=a4a" style=width:350px;height:100px; type=cloudflare width=350>

Thank you

**Comment by jasti** [comment]
@lannka is template ads ready to be shared for Dianomi's use?

**Comment by dianomi-mjs** [comment]
Hi, is there anything we need to do to progress things?

**Comment by dianomi-mjs** [comment]
Looks like I didn't manage to paste our integration code a few comments back ... here is it

<amp-ad class="i-amphtml-layout-fixed i-amphtml-layout-size-defined" data-cf-network=dianomi data-id=3186 height=100 i-amphtml-layout=fixed src="https://www.dianomi.com/smartads.pl?format=a4a" style=width:350px;height:100px; type=cloudflare width=350>

**Comment by dianomi-mjs** [comment]
Looks like I didn't manage to paste the integration code in the original description or it got removed somehow.  Here it is

`<amp-ad class="i-amphtml-layout-fixed i-amphtml-layout-size-defined" data-cf-network=dianomi data-id=3186 height=100 i-amphtml-layout=fixed src="https://www.dianomi.com/smartads.pl?format=a4a" style=width:350px;height:100px; type=cloudflare width=350>`

**Comment by lannka** [comment]
hi @dianomi-mjs , sorry for the delay.

as @jasti mentioned, the other option is to use "template ads" to completely skip the signing. The concept is simple:
1. the ad template needs to be served from AMP cache
2. the ad response will contain the template URL and the data to populate the template.
3. the ad is fully client side rendered

That said, it will require some work at your side to get integrated. Check [here](https://docs.google.com/document/d/1nF1L_1rAn7I5culLX0_6HqQaCWTfmu0nHaYqwMbfnGI/edit#heading=h.ha73jvcfte12) for some more details.

**Comment by dianomi-mjs** [comment]
adding @michael-stevens

**Comment by michael-stevens** [comment]
Hi,

We've tried to create a fast fetch implementation as documented at https://github.com/ampproject/amphtml/blob/master/ads/google/a4a/docs/Network-Impl-Guide.md#creating-an-amphtml-ad-extension-implementation

and I have an amp-ad tag that seems to work on a test page served locally from our https://github.com/dianomi-mjs/amphtml fork. 

I couldn't find any documentation how to handle the validation and signing part so I'm not sure how to handle that.

At the moment this is looking like an easier integration for us than the template option, but it obviously depends on how the signing works. Can you supply any info? Or is it possible to just work with what I've done so far without the signing component?

**Comment by jasti** [comment]
Hi @michael-stevens, unfortunately, signing is only available to entities that a) are independent(e.g. Cloudflare) b) or serve > 20% of the total ads served to AMP pages to maintain trust. (e.g. AdSense)
b) Is a requirement because AMPHTML ads have preferred rendering in AMP pages and signing privileges are available once the network gains a good amount of traction. 

In the mean time, template ads are the right alternative options. Let us know if you have any other questions.

**Comment by dianomi-mjs** [comment]
Hi @jasti - can you tell us who are the independent entities we could use now, seeing as cloudflare are closing down their service?

**Comment by jasti** [comment]
@dianomi-mjs Unfortunately there aren't other independent entities, so template ads is the only option.

**Comment by dianomi-mjs** [comment]
Hi @jasti ok then - so we read that

""Note: at this stage, domains need to be whitelisted manually in AMP Cache to experiment the feature." "
Can we have www.dianomi.com whitelisted in AMP Cache please?

**Comment by jasti** [comment]
to @lannka.

**Comment by lannka** [comment]
@dianomi-mjs sure thing. FYI, it takes roughly a week to get it whitelisted. While waiting, you can start following the instructions for the integration. Let me know if you encounter any issue down the line.

**Comment by lannka** [comment]
@dianomi-mjs forgot to ask, what path do you plan to serve those templates?

for example, it can be: `https://static.dianomi.com/amp/`

**Comment by lannka** [comment]
It will be better if you can have a specific path for all the templates. For the URL you provided, we have to whitelist your entire domain.

**Comment by dianomi-mjs** [comment]
Hi there

I copied your example here

https://www.dianomi.com/amp/amp4ads_0.epl

Is that ok?

Thanks

**Comment by lannka** [comment]
That looks good. I will let you know when the whitelisting takes effect.

Meantime, let me know if you have any issue following the instruction for the integration.

**Comment by lannka** [comment]
ah, sorry shared a wrong link.
the correct one should be: https://lannka.github.io/amp-ad-template/amp4ads_0.html

could you please update?

**Comment by dianomi-mjs** [comment]
Hi

I've updated 

Is that better?

Thanks

**Comment by lannka** [comment]
Yep. The whitelisting is already done. I'll post here once it gets into production.

**Comment by lannka** [comment]
@dianomi-mjs it's now available:

The template is now served from AMP cache:
https://www-dianomi-com.cdn.ampproject.org/ad/s/www.dianomi.com/amp/amp4ads_0.epl

Let me or @glevitzky know if there is any problem you found while integrating template ads.
/cc @keithwrightbos

**Comment by dianomi-mjs** [comment]
Hi, thanks

So based on the documentation I would have expected

<amp-ad-custom width=350 height=100
  src=”https://www.dianomi.com/smartads.pl?format=amp”
  data-request-param-id="3186"
</amp-ad-custom>

to generate a request for the URL given (I wouldn't have expected it to work yet as there is a lot more we need to put in to place)

But instead I am seeing javascript errors, and no request.

I put this on this page underneath our existing ad served via cloudflare

https://amp-kiplinger-com.cdn.ampproject.org/v/amp.kiplinger.com/slideshow/retirement/T006-S001-worst-states-for-retirement-2018.html?usqp=mq331AQECAFYAQ%3D%3D&amp_js_v=0.1

Can you advise?

Cheers

**Comment by dianomi-mjs** [comment]
Oh on further inspection I think we still need to make our own a4a extension so we can use amp-ad-dianomi ?

**Comment by glevitzky** [comment]
@dianomi-mjs It looks like your tag is set up correctly, and you don't need your own a4a extension (unless you want to add additional custom logic to your ad request -- this can be done later once we hammer out some initial set up issues, if you wish).

I _believe_ the issue is that you're not adding `<script async custom-element="amp-ad-custom" src="https://cdn.ampproject.org/v0/amp-ad-custom-0.1.js"></script>` to your page. Apologies that this isn't documented more clearly. Could you try adding this script tag and trying again?

**Comment by dianomi-mjs** [comment]
Thanks - that seems to be getting further now.

I know there's a lot more I need to do, and will continue to follow the documentation provided earlier by @lannka but anything else that would be useful would be appreciated.

**Comment by dianomi-mjs** [comment]
Getting further

I still got errors when I added

AMP-Ad-Template_Extension: amp-mustache
Access-Control-Expose-Headers: AMP-Access-Control-Allow-Source-Origin

I needed to also add:

AMP-Access-Control-Allow-Source-Origin: https://www.dianomi.com

**Comment by glevitzky** [comment]
Noticed another thing. This is the <amp-ad-custom> element as copied from your test page:

```HTML
<amp-ad-custom width="350" height="100" src="https://www.dianomi.com/smartads.pl?format=amp” data-request-param-id=" 3186"="" class="i-amphtml-element i-amphtml-layout-fixed i-amphtml-layout-size-defined i-amphtml-layout" style="width: 350px; height: 100px;">
</amp-ad-custom>
```

It looks like the closing quote on the src attribute isn't a standard quote character. Once I fixed that, I see an ad request go out that looks sensible to me. 

(There's also an extraneous `=""` following the data-request-param-id attribute, though I don't think it's causing any harm.)

**Comment by dianomi-mjs** [comment]
Hi, Sorry about that - I wrote an updated test here where I managed to get much further

https://www.dianomi.com/mjs/amp/test.epl

The amp request is now being made, I'm trying to work out why the template isn't being loaded but going through the instructions.

**Comment by dianomi-mjs** [comment]
Hi, @glevitzky I wonder if you could take a look at the page we have now

https://www.dianomi.com/mjs/amp/test.epl

On chrome I seeing an "Invalid recovery mode!" error

on the chromebook I don't see an error but it looks like it's posting an error to https://amp-error-reporting.appspot.com/r

{v: "011901302225410", noAmp: "0", m: "Invalid recovery mode!", a: "0", ex: "0", dw: "0", rt: "1p",…}
a: "0"
ae: ""
bt: "production"
ca: "0"
dw: "0"
el: "u"
ex: "0"
exps: "amp-consent=1,chunked-amp=1,hidden-mutation-observer=1,expAdsenseCanonical=0,canary=0,expAdsenseUnconditionedCanonical=0,amp-story-v1=1,ad-type-custom=1,linker-form=1,no-sync-xhr-in-ads=1,a4aProfilingRate=0,version-locking=1,scroll-height-bounce=0,amp-sidebar toolbar=1,amp-auto-ads-adsense-holdout=0,amp-list-resizable-children=1,as-use-attr-for-format=0,amp-auto-ads-adsense-responsive=0,font-display-swap=1,amp-ima-video=1,amp-playbuzz=1,amp-auto-ads=1,pump-early-frame=1,amp-date-picker=1,inabox-rov=1,amp-access-iframe=1,a4aFastFetchAdSenseLaunched=0,amp-list-viewport-resize=1,doubleclickSraExp=0,amp-apester-media=1,doubleclickSraReportExcludedBlock=0,amp-live-list-sorting=1,linker-meta-opt-in=1,ampdoc-closest=0,amp-story-responsive-units=1,scroll-height-minheight=0,user-error-reporting=1,expAdsenseA4A=0,a4aFastFetchDoubleclickLaunched=0,no-initial-intersection=1,sandbox-ads=1"
fr: ""
jse: "Chrome"
m: "Invalid recovery mode!"
noAmp: "0"
r: ""
rt: "1p"
rvu: "https://www.dianomi.com/mjs/amp/test.epl"
s: "amp-ad-network-base: Invalid recovery mode!↵    at Ia (https://cdn.ampproject.org/v0.js:12:169)↵    at Fa.f.Ea (https://cdn.ampproject.org/v0.js:7:197)↵    at Fa.f.error (https://cdn.ampproject.org/v0.js:7:307)↵    at Ya (https://cdn.ampproject.org/v0/amp-ad-custom-0.1.js:23:171)↵    at https://cdn.ampproject.org/v0/amp-ad-custom-0.1.js:20:147"
v: "011901302225410"

**Comment by glevitzky** [comment]
Looks like the actual error message is being swallowed up somewhere, but stepping through the debugger, I see: `"XHR Failed fetching (https://www.dianomi.com/...): Failed to fetch​​​"`. Not sure yet if the issue is on our side or not, but thought I'd post this in case it rings any bells for you while I keep debugging.

**Comment by michael-stevens** [comment]
I've possibly? made some progress by running a local copy of amphtml, which seems to give better errors, we had some dodgy CORS headers, we now see the error "Validator never registered!".

**Comment by glevitzky** [comment]
Have you updated your test page? I'm seeing the old error still.

**Comment by michael-stevens** [comment]
That's not updated yet, I'll try to get that updated.

**Comment by dianomi-mjs** [comment]
Hi

We've updated our headers

So I have this page the firebolt and template ad

https://www.dianomioffers.co.uk/mjs/amp/test.epl

and this one with just the template ad

https://www.dianomioffers.co.uk/mjs/amp/newonly.epl

We're seeing "template-impl.js:241 Invalid recovery mode!"

but can't work out what that means. 

Thanks for any help!

**Comment by glevitzky** [comment]
This is our mistake for not documenting this, but you'll need to add an additional HTTP response:

`AMP-Ad-Response-Type: template`.

This should be in the same response that contains the `amp-ad-template-extension: amp-mustache` header.

**Comment by glevitzky** [comment]
In principle, this shouldn't be needed, and I'm currently working on a fix that will make this optional, but for current testing purposes, adding this header should get you passed the current block.

**Comment by glevitzky** [comment]
Having bypassed that check, I still see further issues in that your actual template isn't being requested. Currently investigating why.

**Comment by dianomi-mjs** [comment]
No problem. We've added the header now.

On https://www.dianomi.com/mjs/amp/newonly.epl I don't see any errors (but no template as I think you've also observed)

On https://www.dianomioffers.co.uk/mjs/amp/newonly.epl (same content on different domain to the JSON) we still get the invalid recovery mode.

We're investigating on our side too.

**Comment by glevitzky** [comment]
It looks like for some reason one of the headers is being dropped. The check here: https://github.com/ampproject/amphtml/blob/master/extensions/amp-a4a/0.1/template-validator.js#L65 appears to be failing, even though I can clearly see that the header is present in the HTTP response. Still debugging...

**Comment by glevitzky** [comment]
Actually, that could be due to my local test setup. Disregard the above.

**Comment by glevitzky** [comment]
It looks like in the first link the template is now being fetched correctly. However, it's missing the `"ampRuntimeUtf16CharOffsets" ` field in the metadata object, which is causing it to be flagged as invalid. This field needs to specify the first and and last characters of the metadata object script tag. The form should be:

```json
<script amp-ad-metadata type=application/json>{
   "customElementExtensions" : [ "amp-analytics" ],
   "extensions" : [
      {
         "ampRuntimeUtf16CharOffsets": [positionOfFirstCharacterOfScriptTag, positionOfLastCharacterOfScriptTag],
         "custom-element" : "amp-analytics",
         "src" : "https://cdn.ampproject.org/v0/amp-analytics-0.1.js"
      }
   ],
   "images" : [
      "https://validator.ampproject.org/logo-blue.svg",
      "https://validator.ampproject.org/logo-blue.svg"
   ]
}
</script>
```

This field isn't particular to template ads, but is expected of all AMP creatives.

**Comment by dianomi-mjs** [comment]
Thanks - so I presume we need to add a line to our original file https://www.dianomi.com/amp/amp4ads_2.epl (which is a small modification of the file @lannka sent).

I'll see if I can work out what this entails.

**Comment by dianomi-mjs** [comment]
actually is this something we should be adding or should it be being added for us?

**Comment by glevitzky** [comment]
My mistake. This should be auto-injected for you. We're currently investigating why the string wasn't properly constructed. I'll update you as soon as I know what's going on.

**Comment by dianomi-mjs** [comment]
No problem - please let us know when there's anything we can do.

**Comment by lannka** [comment]
hi @dianomi-mjs we've located the bug and will keep you posted for any progress. thanks for your patience.

**Comment by dianomi-mjs** [comment]
Thanks for the update!

**Comment by dianomi-mjs** [comment]
Hi @lannka anything for us to do at the moment (other than be patient)?

Thank you

**Comment by dianomi-mjs** [comment]
Hi @lannka @glevitzky any update for us?

Thanks

**Comment by lannka** [comment]
@dianomi-mjs sorry we're waiting for Google Cache to fix the issue. Will post here once we get an ETA for it.

**Comment by dianomi-mjs** [comment]
Thanks - is there any alternative? At the moment we're at the whim of cloudflare keeping their service running for us ...

**Comment by lannka** [comment]
Unfortunately not. /cc @jasti @Gregable for prioritization.

**Comment by dianomi-mjs** [comment]
Hi @jasti @Gregable sorry to be a pain, can we get any timeframe on when the issue will be fixed and we can start serving template ads?

**Comment by glevitzky** [comment]
Hi @dianomi-mjs. Apologies for the continued delays. We are working on deploying a client-side fix targeted for next week's canary release, which should unblock us for our purposes here.

**Comment by lannka** [comment]
> next week's canary release

which is next Tuesday. @dianomi-mjs that should unblock your development.

**Comment by dianomi-mjs** [comment]
Hi @lannka @glevitzky did the fix make it in to the canary release?

I tried enabling the dev channel on https://cdn.ampproject.org/experiments.html but couldn't see a change.

Thanks

**Comment by glevitzky** [comment]
It doesn't look like a canary has been released this week yet. I will put an update here when that happens.

**Comment by glevitzky** [comment]
Hi @dianomi-mjs, I believe the change should be in the latest canary. Please give it a try and let me know how things are.

**Comment by dianomi-mjs** [comment]
Hi @glevitzky do I enabled canary by selecting 'AMP Dev Channel'?  If so then 

https://www.dianomioffers.co.uk/mjs/amp/test.epl still gives me an invalid recovery mode
and
https://www.dianomi.com/mjs/amp/test.epl gives me new and different errors and an unfilled template (at least on linux, oddly on my chromebook I see no errors but also no template):


```
Refused to load the script 'https://cdn.ampproject.org/rtv/011903141753530/amp4ads-v0.js.br' because it violates the following Content Security Policy directive: "script-src 'none'". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.

v0.js:46 Uncaught TypeError: Cannot read property 'create' of undefined
    at new a.(anonymous function) (https://cdn.ampproject.org/v0.js:46:138)
    at new Ac (document-register-element.patched.js:302)
    at document-register-element.patched.js:291
    at Qc (document-register-element.patched.js:299)
    at Sh (amp-img.js:164)
    at Kh.f.installExtensionsInChildWindow (custom-element-registry.js:117)
    at friendly-iframe-embed.js:212
```


Thank you

**Comment by glevitzky** [comment]
Hi @dianomi-mjs, yes, canary is enabled by opting into "AMP Dev Channel".

It looks like on https://www.dianomi.com/mjs/amp/test.epl, you just need to give the ad slot more height. Moreover, all template variables need to be within the <template> tag, if they are to be filled correctly. Otherwise, things look like they're working as intended here. The error you posted is a known, but unrelated issue.

Still investigating the issue on https://www.dianomioffers.co.uk/mjs/amp/test.epl. Will update you as soon as I can.

**Comment by dianomi-mjs** [comment]
Hi @glevitzky Great - thanks, that gives me something to go on.

Any change on the original resource takes a while to permeate through (I've not worked out how long yet but it's not as if the cache gets updated after each request despite Cache-Control headers being no-cache, no-store, private). Which makes development a bit tricky. Any ways to easily control this?

**Comment by dianomi-mjs** [comment]
I've managed to get the ad working on https://www.dianomi.com/mjs/amp/test.epl, so looking good in this context!

**Comment by glevitzky** [comment]
Glad to hear of your success!

On https://www.dianomioffers.co.uk/mjs/amp/test.epl, for some reason we are unable to read the "amp-ad-response-type" header, which is causing the observed error. I've compared the HTTP response from that page against the one from https://www.dianomi.com/mjs/amp/test.epl, and the only difference is that in the latter you use `access-control-allow-origin: *` while in the former you use `access-control-allow-origin: https://www.dianomioffers.co.uk`. I can't tell if that's the reason, and I don't really see why it would be, but it appears to be the only difference between the two responses. Maybe it has to do with the fact that you're requesting the resource from `https://www.dianomi.com/smartads.pl?...`?

**Comment by dianomi-mjs** [comment]
Hi @glevitzky 

Thank you for that - I used the excellent [requestly plugin](https://chrome.google.com/webstore/detail/requestly-redirect-url-mo/mdnleldcmiljblolnjhpnblkcekpdkpa?hl=en) modify the response header to access-control-allow-origin: *. This unfortunately gives me a whole host of console errors starting with

Access to fetch at 'https://www.dianomi.com/smartads.pl?format=amp&id=3186&__amp_source_origin=https%3A%2F%2Fwww.dianomioffers.co.uk' from origin 'https://www.dianomioffers.co.uk' has been blocked by CORS policy: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'.

I have to admit I really struggle with CORS - but presumably the issue here really is that for some reason I'm unable to successfully request the resource from a different domain, which is something we really need to do to serve ads of course! When the resource is on the same domain the Access-Control-Allow-Origin headers are irrelevant I believe.

**Comment by glevitzky** [comment]
Okay, made a little progress. You want to add `amp-ad-response-type` to `access-control-expose-headers`. Verified using requestly that this eliminates the "Invalid Recovery Mode" error, but there's still some further issues. Investigating now.

**Comment by glevitzky** [comment]
Think I got it. You need to add both `amp-ad-template-extension` and `amp-ad-response-type` to `access-control-expose-headers`. So the full header should look like: `access-control-expose-headers: AMP-Access-Control-Allow-Source-Origin, AMP-Access-Control-Allow-Source-Origin,amp-ad-response-type,AMP-Ad-Template-Extension` (case-insensitve).

**Comment by dianomi-mjs** [comment]
Thanks @glevitzky 

https://www.dianomioffers.co.uk/mjs/amp/test.epl

now seems to work!

**Comment by jasti** [comment]
@dianomi-mjs congrats!
Have you seen any latency differences between the two?

**Comment by dianomi-mjs** [comment]
@jasti Nothing obvious - but both are noticeably faster than the cloudflare served version

**Comment by dianomi-mjs** [comment]
Hi, this seems to work fine without the dev channel I believe ... so I think we are almost ready to roll this out. Would be nice to have the javascript errors resolved though, some people do get upset by that.

Thanks for the help so far!

**Comment by jasti** [comment]
@dianomi-mjs if you are interested, if we'd be happy to showcase you as an example using template ads at AMP conf on Apr 17th in Tokyo: https://www.ampproject.org/amp-conf/
Let me know what you think and an email address to reach out to and I can send over more details. Thanks!

**Comment by dianomi-mjs** [comment]
Hi - sure thing.  You can reach me at

michael.smith@dianomi.com

It all seems to work fine for our other units now. So we just need to get our publishers to update their code. Hopefully we can roll out some brand new units too.

**Comment by dianomi-mjs** [comment]
@jasti just wanted to make sure you saw the last comment.

**Comment by dianomi-mjs** [comment]
Hi

It's working well and we're trying to get a few publishers to deploy.

One wants to run our AMP ads through DFP - is this possible or does the integration need to be direct?

Thanks

**Comment by lannka** [comment]
@dianomi-mjs hi glad to see it finally worked. Thanks for your patience and great work!

At this moment DFP does not support it. So it has to use your tag.

**Comment by dianomi-mjs** [comment]
Hi

Sorry to update with further questions - in order for publishers to deploy template ads on their site, do they need to add the following to their header?

<script async custom-element="amp-ad-custom" src="https://cdn.ampproject.org/v0/amp-ad-custom-0.1.js"></script>

If so do they need to specify the version or is there a better way?  Or maybe this will become standard rather than an extensions?

Also I am still seeing quite a lot of console errors / warnings which I don't think it my doing (and might concern some publishers)

eg https://www.dianomi.com/mjs/amp/ccn.epl

Thanks
Michael

**Comment by lannka** [comment]
@dianomi-mjs hi Michael, yes like any other AMP components, pub needs to include the amp-ad-custom script in the head with its version number.

The console errors are known issues #22264 #17997 , it's safe to ignore them.

**Comment by dianomi-mjs** [comment]
Hi @lannka 

Thanks - our first publisher has told us that the page doesn't validate when they include the tag, and the ads are blank.

Here's my version

https://www.dianomi.com/mjs/amp/ccn.epl

and indeed the validator isn't happy

https://validator.ampproject.org/#url=https%3A%2F%2Fwww.dianomi.com%2Fmjs%2Famp%2Fccn.epl

and also

https://search.google.com/test/amp?id=dXcA4ulve9eyXahnlK3tJg

Speculating a bit that there's a list of valid javascript that can be loaded in, and amp-ad-custom isn't on that list yet?

Thanks

**Comment by lannka** [comment]
That was indeed an oversight. Will fix ASAP: #22280 
sorry for the back-n-forth

**Comment by lannka** [comment]
FYI, the validator fix is in. It will be out in production next week.

**Comment by dianomi-mjs** [comment]
Thanks @lannka I can see that the google api now says the page is valid :)

https://search.google.com/test/amp?id=2zNfQftfxvA93CKj84I3JQ

Not quite sure exactly how things should work for a publisher but keen to be sure it'll work for them before I go back to them. I think I'd need to get the page served by the amp cache.

When I request

curl -i -s -k -X POST -H "Content-Type: application/json" -H "X-Goog-Api-Key: [mykey]" -d "{urls: ['https://www.dianomi.com/mjs/amp/ccn.epl']}" "https://acceleratedmobilepageurl.googleapis.com/v1/ampUrls:batchGet"

I always get

{
  "urlErrors": [
    {
      "errorCode": "NO_AMP_URL",
      "errorMessage": "No AMP URL for the request URL.",
      "originalUrl": "https://www.dianomi.com/mjs/amp/ccn.epl"
    }
  ]
}

likewise I *think* this ought to work

curl 'https://www-dianomi-com.cdn.ampproject.org/c/s/www.dianomi.com/mjs/amp/ccn.epl'

but it directs me to my hosted version.

Thanks

**Comment by lannka** [comment]
@dianomi-mjs the change is not fully propagated to everywhere yet. will update the thread once it's fully working.
thanks for your patience again

**Comment by lannka** [comment]
Seems the link is working now. Pls check.

**Comment by dianomi-mjs** [comment]
Thanks - I can now get the page served via the AMP cache

https://www-dianomi-com.cdn.ampproject.org/c/s/www.dianomi.com/mjs/amp/ccn.epl

I am seeing lots of access control errors now though, which I'm trying to get my head round:

error.js:198 Returned AMP-Access-Control-Allow-Source-Origin is not equal to the current: https://www-dianomi-com.cdn.ampproject.org vs https://www.dianomi.com

I think my cgi-scripts are doing the right thing

Request:
Origin: https://www-dianomi-com.cdn.ampproject.org

Response:
access-control-allow-origin: https://www-dianomi-com.cdn.ampproject.org

Unless I need to somehow make the response be https://www.dianomi.com?

**Comment by lannka** [comment]
yep, you need to do  `https://www.dianomi.com`

**Comment by dianomi-mjs** [comment]
Thanks - sorted that

Last (hopefully!) thing - we're struggling to get responsive ads working - as per this jsfiddle (shows comparison with the cloudflare ad system)

https://jsfiddle.net/raj177/03zeqbw8/3/

Should this work or are there features of amp-ad not available in amp-ad-custom?

**Comment by lannka** [comment]
@dianomi-mjs yep, we have a TODO for it. Is it a blocking issue for you? If so, we can prioritize it.
added a tracker for this: #22465

**Comment by dianomi-mjs** [comment]
no desperate hurry I think ... we are going over all our existing AMP units and asking publishers to transition. This is the only affected unit we've found so far. Would obviously be good to fix but as long as cloudflare keep their product going, not urgent.

**Comment by lannka** [comment]
@dianomi-mjs the fix PR is already in. you can expect it to be in today's canary build. Let me know if there is anything else. All those are super helpful.

**Comment by dianomi-mjs** [comment]
Hi @lannka , I enabled the dev channel but don't see a change. Did the fix get deployed?  Thank you

**Comment by lannka** [comment]
It's already in canary. It will hit prod next Tuesday.

**Before**
<img width="375" alt="Screen Shot 2019-05-31 at 9 43 35 AM" src="https://user-images.githubusercontent.com/8496897/58721063-c119df80-8388-11e9-806c-41c771293b2a.png">

**After**
<img width="726" alt="Screen Shot 2019-05-31 at 9 44 22 AM" src="https://user-images.githubusercontent.com/8496897/58721069-c414d000-8388-11e9-8332-0707ecacf1f8.png">

**Comment by dianomi-mjs** [comment]
Hi there

We are seeing some problems in our own metrics system which I've been investigating. We are counting the number of requests made for our ads and the numbers are an awful lot higher than in the previous system.

I've been trying to work out why. If you load this page and look at the network tab you can see many many requests coming though, especially if you for example drag the line that resizes the main window vs developer window. It seems to generate dozens of requests.

Is there a way of us ensuring that there's just one request each time?

Thanks

**Comment by dianomi-mjs** [comment]
forgot the URL: 

https://www.dianomi.com/mjs/amp/one.epl?smartad_id=5001

**Comment by lannka** [comment]
thanks for reporting. issue tracked here: #23094 
will have someone look at it ASAP.
